pandas/tests/io/formats/test_to_csv.py

```diff
@@ -351,3 +352,15 @@ def test_to_csv_compression(self, compression_only,
             result = pd.read_csv(path, index_col=0,
                                  compression=read_compression)
             tm.assert_frame_equal(result, df)
+
+    def test_compression_defaults_to_infer(tmpdir):
+        """
+        Test that to_csv defaults to inferring compression from paths.
+        https://github.com/pandas-dev/pandas/pull/22011
+        """
+        df = DataFrame({"A": [1]})
```

Comment:
While I suppose this "works" it is definitely focused on `gzip` and doesn't make any assertions about how the behavior works across a combination of `infer`, `None` and an explicit compression.

We have a fixture called `compression` in `conftest.py` that covers the various compression options - is there a way to parametrize that as part of this test and make more comprehensive assertions about the potential combinations of behavior?

Reply:
The possible values that compression can take for `to_csv` are already tested. The purpose of this test is simply to test that `compression` is defaulting to `'infer'`. I kept the test simple so that it ONLY tests that the default is infer and won't break or malfunction should other aspects of the API change. The test will break if the default for compression is put back to None (and hopefully not for other reasons).

I think it may make sense to make a similar test for `to_json`, but don't see how testing `compression='infer'` for all possible compression extensions is within scope of this PR as this PR just changes the default and is not creating the infer option. Is the worry that there is currently inadequate testing?

Reply:
> it should be that much more work to leverage the existing compression fixture for this test to increase coverage

In https://github.com/pandas-dev/pandas/pull/22011/commits/abd19e3b02c77e1160d460ab50b34d4642d568c2, I modified an existing parametrized test to look for compression by default for paths where inference should occur. This actually caught an issue (we hadn't switched default for `Series.to_csv` -- fixed in https://github.com/pandas-dev/pandas/pull/22011/commits/2f670fedbe1694e64096c2a775d8e83799f04036).

Should I delete the gzip test? Note the parametrized test doesn't test that the _right_ compression is occurring, just that a compression is occurring.

Reply:
Instead of going about it in this fashion can you not do a round trip with `to_csv` and `read_csv` using infer with a file extension for the former and the parametrized value as an argument for `compression` in the latter? So in pseudo-code:

```python
with tm.ensure_clean('compressed.csv.{}'.format(compression_only)) as path:
    df.to_csv(path)
    result = pd.read_csv(path, compression=compression_only)

tm.assert_frame_equal(result, df)
```

Reply:
I'm not a big fan of the roundtrip approach here, since it never actually tests that the file is compressed on disk. Given that the to_* and read_* methods rely on much of the same compression infrastructure, I think it's possible to modify the code such that all compression gets disabled and the roudtrip works perfectly. Now hopefully there are enough other tests to catch such a situation.